### PR TITLE
Return table or view when listing for tables with fully qualified name

### PR DESCRIPTION
**Pull request by wendigo**
Fixes https://github.com/prestosql/presto/issues/6239

### CODE REVIEW

**Comment by kokosing** [comment]
For testing you can use something similar to `io.prestosql.metadata.TestInformationSchemaMetadata` or `io.prestosql.connector.system.TestSystemConnector`

**Comment by losipiuk** [comment]
> For testing you can use something similar to `io.prestosql.metadata.TestInformationSchemaMetadata` or `io.prestosql.connector.system.TestSystemConnector`

Look for usages of `MockConnectorFactory`.

**Review by losipiuk — COMMENTED** [review]
Code LGTM. Please add some test coverage.

**Comment by wendigo** [comment]
AC @losipiuk

File: presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java [hunk 1]
```diff
@@ -576,6 +576,15 @@ public ColumnMetadata getColumnMetadata(Session session, TableHandle tableHandle
         return ImmutableList.copyOf(tables);
     }
 
+    private Optional<QualifiedObjectName> getTableOrView(Session session, QualifiedObjectName name)
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:579]: i think this should be

```
boolean isExistingRelation(Session session, QualifiedObjectName name)
```

**Comment by findepi** [comment]
We should also fix `io.prestosql.plugin.hive.HiveMetadata#getTableHandle` not to return a table handle for a view.
This is the thing that allowed https://github.com/prestosql/presto/issues/6239 be uncovered until recently.

**Review by losipiuk — APPROVED** [review]

File: presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java (hunk 1 above)
Comment (wendigo) [reply presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java]: Done

File: presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java [hunk 2]
```diff
@@ -332,23 +332,19 @@ public boolean schemaExists(Session session, CatalogSchemaName schema)
             return Optional.empty();
         }
 
-        Optional<CatalogMetadata> catalog = getOptionalCatalogMetadata(session, table.getCatalogName());
-        if (catalog.isPresent()) {
-            CatalogMetadata catalogMetadata = catalog.get();
+        return getOptionalCatalogMetadata(session, table.getCatalogName()).flatMap(catalogMetadata -> {
```
Comment (losipiuk) [inline presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:335]: I would not call it simplification :p

**Review by losipiuk — APPROVED** [review]

File: presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java [hunk 3]
```diff
@@ -73,6 +73,18 @@ default ConnectorTableHandle getTableHandle(ConnectorSession session, SchemaTabl
         return null;
     }
 
+    /**
+     * Return true if table or view with specified name exists.
+     */
+    default boolean isExistingRelation(ConnectorSession session, SchemaTableName objectName)
```
Comment (losipiuk) [inline presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:79]: This is an SPI change now. I am not sure we want this.
@findepi, @martint ?

Comment (wendigo) [reply presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:79]: I'll remove that.

File: presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java [hunk 4]
```diff
@@ -576,6 +577,13 @@ public ColumnMetadata getColumnMetadata(Session session, TableHandle tableHandle
         return ImmutableList.copyOf(tables);
     }
 
+    private boolean isExistingRelation(Session session, QualifiedObjectName name)
+    {
+        return getTableHandle(session, name).map(handle -> true)
```
Comment (losipiuk) [inline presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:578]: Personally I would prefer just two `if`s here :)

**Review by losipiuk — APPROVED** [review]

**Comment by losipiuk** [comment]
Mind the red X-es

**Comment by wendigo** [comment]
@losipiuk caused by https://github.com/prestosql/presto/pull/6259. I've rebased that change so it should pass checkstyle as well

**Comment by wendigo** [comment]
presto-memory is failing (unrelated). I'm gonna link an issue.

**Pull request merged**